### DISCUSSION TITLE
how to deal with files vweb

### DISCUSSION THREAD

**Discussion by despiegk** [vlang/v#11809] [Ideas]
Hi,

Have been reading the source code of vweb and I don't find a convenient way how to return files

remarl:
- should not be loaded in memory before sending to client

would be nice there is just a method like return app.file_send(path) ...

is there any way how this can be done?

many thanks

Kristof

**Comment by despiegk** [comment]
related question is there a way how to do something like ['/:users/.*'] to capture the full path behind users?

**Comment by medvednikov** [comment]
You mean something like this?

https://stackoverflow.com/questions/24116147/how-to-download-file-in-browser-from-go-server

**Comment by axetroy** [comment]
I think vweb is a failed design, and only basic HTTP is needed in the standard library like other languages does .

vlib should provide the low-level interface

Functions such as framework/routing should be provided by third-party packages. Implementing the functions of the framework in the standard library will definitely cause a burden in the end.

**Comment by medvednikov** [comment]
V is a batteries included language.

We don't want the developer to chose across 20 different frameworks.

Writing web services is a very common task. vweb for V will be like Rails for Ruby.

**Comment by dumblob** [comment]
@axetroy I think you're raising a good point - especially the question regarding future. In my opinion HTTP is quite stable and the number of changes to related RFCs is probably comparable to the frequency with which *V* wants to be developed in the next 10 or 15 years (if you disagree, please list your arguments - I think they'll be important to discuss).

On the other hand web technologies are an enormously moving target. So I do agree, that providing a web **framework** might be problematic (to keep up the tempo **while** not becoming bloated due to expanding API).

If you have any specific points you think `vweb` should do differently in terms of lower risk of deprecation in the coming years (e.g. you see a proliferation of a different technique for some `vweb` functionality which seems to have the potential to replace what's currently being used in `vweb`), **please speak up**, so that we can either change it in `vweb` now or maybe throw that part of functionality completely away (as you've indicated in your comment suggesting making `vweb` lower-level) if we see the risk of fluctuating (bloating) API is too high.

**Comment by medvednikov** [comment]
@dumblob what has changed in web servers in the last 15 years? Lots of changes in front-end frameworks, but Rails for example still pretty much works the same way.

**Comment by dumblob** [comment]
>what has changed in web servers in the last 15 years?

The requirement to handle orders of magnitude more simultaneous connections and at the same time have much lower latencies (because of streaming of all kind becoming ubiquitous - see e.g. [nng Scalability Protocols](https://staysail.github.io/nng_presentation/nng_presentation.html ) as an alternative showing "the direction" which is being then replicated using HTTP in web servers).

>Lots of changes in front-end frameworks, but Rails for example still pretty much works the same way.

That's true. But "nobody" uses Rails nowadays if you look at statistics - because it's far too slow for modern workloads (e.g. streaming of all kinds of information, push notifications, low latencies, etc.).

------------------

Btw. "streaming of all kind" means not only streaming of an integer to the client and back, it's becoming difficult for the web server if there are bigger chunks to be streamed - e.g. whole complex list of thousands of products for sale or nowadays more and more ubiquitous the whole page/DOM. And I have to admit, that I don't think `vweb` has the best API for e.g. whole page/DOM streaming (because `vweb` API focuses on the individual "samples" - the rendered pages - instread of on the whole indefinite stream of samples/changes).

**Comment by JalonSolov** [comment]
There is nothing stopping anyone from using an alternative to vweb.  For example, vex has been around a while.

vweb is there to get someone started, and it handles many of the more common features people want in web servers.
